### DISCUSSION TITLE
camera adjuster fails

### DISCUSSION THREAD

**Discussion by VeeranjaneyuluToka** [OpenStitching/stitching#235] [Q&A]
Hi,

Nice and very detailed implementation.

I tried your stitched detailed notebook on our own set of images (no.of images are 3). but it fails with below error

---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
/tmp/ipykernel_2447557/2344847478.py in <module>
      8 
      9 cameras = camera_estimator.estimate(features, matches)
---> 10 cameras = camera_adjuster.adjust(features, matches, cameras)
     11 cameras = wave_corrector.correct(cameras)

/mmdetection3d/tools/marine/stitching/camera_adjuster.py in adjust(self, features, pairwise_matches, estimated_cameras)
     44 
     45     def adjust(self, features, pairwise_matches, estimated_cameras):
---> 46         b, cameras = self.adjuster.apply(features, pairwise_matches, estimated_cameras)
     47         if not b:
     48             raise StitchingError("Camera parameters adjusting failed.")

error: OpenCV(4.9.0) /io/opencv/modules/calib3d/src/compat_ptsetreg.cpp:125: error: (-215:Assertion failed) !err.empty() in function 'update'

I tried with all the camera adjusters, it works with no but none of other options works. I am not getting correct outuput with no adjuster. Could you please let me know what could be the reason and possible ways to analyse further on this issue.

Thanks,
Veeru.

**Comment by lukasalexanderweber** [comment]
Hey, how do you aquire your Images? Have a look on https://github.com/OpenStitching/stitching/discussions/135#discussioncomment-6816177

**Reply by lukasalexanderweber** [reply]
Oh I misread that you tried the original images from the notebook

**Reply by VeeranjaneyuluToka** [reply]
No, i am not trying with original images. I am trying with my own images which are generated with 3 cameras. The middle camera is positioned a bit higher than other two cameras.

**Reply by lukasalexanderweber** [reply]
https://stackoverflow.com/questions/78852689/slight-misalignment-issue-on-stitching-images-together-in-grid-like-manner/78855170#78855170

This setup will probably not work

**Reply by HuanhuanZhao08** [reply]
I have the same problem, my image are microscopy images. The features, matches and camera estimates are not empty, what could be the problem?

**Reply by lukasalexanderweber** [reply]
Microscopy images should be stitched in affine mode

**Reply by HuanhuanZhao08** [reply]
I changed the feature matching to affine mode, I still have that error. Feature matching looks fine. It seems this error occur when the confidence score is lower than 1, not sure what the score exactly indicate though.
![Screenshot from 2025-01-16 11-44-36](https://github.com/user-attachments/assets/8ee0ad7f-402a-44a8-bd15-dc946457db39)

![image](https://github.com/user-attachments/assets/b1e50100-abff-45da-825e-a22512187be6)
![image](https://github.com/user-attachments/assets/1a959e3b-0cf6-42df-a946-6200e372242c)

**Reply by lukasalexanderweber** [reply]
Please try the complete affine example with Budapest map

**Reply by HuanhuanZhao08** [reply]
Is the budapest map a dataset?

**Reply by lukasalexanderweber** [reply]
https://github.com/OpenStitching/stitching_tutorial/blob/master/docs/Stitching%20Tutorial.md#affine-stitcher

**Reply by HuanhuanZhao08** [reply]
Thanks, affine works. It is faster at the camera adjustor step and the threshold at the subsetter step can be lower to 0.5.

**Comment by VeeranjaneyuluToka** [comment]
I could get stitched image with 3 images but the output looks like as follows
![image](https://github.com/user-attachments/assets/6ac0bf06-d893-4fed-8b1f-c27db1f82710)

Detection: SIFT
Matcher: Homography
Adjuster: No
Estimator: Homography
wave_corrector: No
wrapper: Sperical
cropper=True
SeamFinder = dp_color
exposure_compensator = gain_blocks
blender=multiband

Do you think changing any of the above settings would give me better output? I also think as i am not applying bundleadjustment, this is how i am getting output. Let me know your comments?
